fog/ledger/enclave/impl/src/oblivious_utils.rs

```diff
@@ -0,0 +1,454 @@
+                key_image_result_code: KeyImageResultCode::KeyImageError as u32,
+            },
+        ];
+
+        let result = collate(client_queries, shard_results);
+
+        let all_images_found = result.iter().all(|key_image_result| {
+            key_image_result.key_image_result_code == KeyImageResultCode::Spent as u32
+        });
+
+        assert!(
+            all_images_found,
+            "{}",
+            "all images are present but some were not collated"
+        );
+
+        // No results available
+        let client_queries = vec![KeyImageQuery {
+            key_image: 1.into(),
+            start_block: 0,
+        }];
+
+        let shard_results = vec![];
+
+        let result = collate(client_queries, shard_results);
+
+        let all_images_found = result.iter().all(|key_image_result| {
+            key_image_result.key_image_result_code == KeyImageResultCode::Spent as u32
+        });
+
+        assert!(
+            !all_images_found,
+            "{}",
+            "all images show results but no result provided"
+        );
+
+        // Result is error
+        let client_queries = vec![KeyImageQuery {
+            key_image: 1.into(),
+            start_block: 0,
+        }];
+
+        let shard_results = vec![KeyImageResult {
+            key_image: 1.into(),
+            spent_at: 1,
+            timestamp: 123,
+            timestamp_result_code: TimestampResultCode::TimestampFound as u32,
+            key_image_result_code: KeyImageResultCode::KeyImageError as u32,
+        }];
+
+        let result = collate(client_queries, shard_results);
+
+        let all_images_error = result.iter().all(|key_image_result| {
+            key_image_result.key_image_result_code == KeyImageResultCode::KeyImageError as u32
+        });
+
+        assert!(
+            all_images_error,
+            "{}",
+            "an image reported no error despite an error result"
+        );
+
+        // Only some queries answered
+        let client_queries = vec![
+            KeyImageQuery {
+                key_image: 1.into(),
+                start_block: 0,
+            },
+            KeyImageQuery {
+                key_image: 2.into(),
+                start_block: 0,
+            },
+        ];
+
+        let shard_results = vec![KeyImageResult {
+            key_image: 1.into(),
+            spent_at: 1,
+            timestamp: 123,
+            timestamp_result_code: TimestampResultCode::TimestampFound as u32,
+            key_image_result_code: KeyImageResultCode::Spent as u32,
+        }];
+
+        let result = collate(client_queries, shard_results);
+
+        let all_images_found = result.iter().all(|key_image_result| {
+            key_image_result.key_image_result_code == KeyImageResultCode::Spent as u32
+        });
+
+        assert!(
+            !all_images_found,
+            "{}",
+            "all images show results but some had no result provided"
+        );
```

Comment:
:wrench: Similar to the not found case above this should be asserting that 1 == spent and 2 == not spent or the default